src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs

```diff
@@ -250,28 +252,43 @@ private static void SetFatalResponse(RequestContext context, int status)
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            _stopping = true;
-            // Wait for active requests to drain
-            if (_outstandingRequests > 0)
+            if (Interlocked.Exchange(ref _stopping, 1) == 1)
             {
-                LogHelper.LogInfo(_logger, "Stopping, waiting for " + _outstandingRequests + " request(s) to drain.");
+                return _shutdownSignal.Task;
+            }
 
-                var waitForStop = new TaskCompletionSource<object>();
-                cancellationToken.Register(() =>
+            try
+            {
+                // Wait for active requests to drain
+                if (_outstandingRequests > 0)
                 {
-                    LogHelper.LogInfo(_logger, "Timed out, terminating " + _outstandingRequests + " request(s).");
-                    waitForStop.TrySetResult(0);
-                });
+                    LogHelper.LogInfo(_logger, "Stopping, waiting for " + _outstandingRequests + " request(s) to drain.");
 
-                return Task.WhenAny(_shutdownSignal.Task, waitForStop.Task);
+                    cancellationToken.Register(() =>
+                    {
+                        LogHelper.LogInfo(_logger, "Timed out, terminating " + _outstandingRequests + " request(s).");
+                        _shutdownSignal.TrySetResult(null);
```

Comment:
Nevermind, this is only the first CT. CTs from all other callers are ignored. That seems even worse.